use_cluster.py

```diff
@@ -0,0 +1,361 @@
+
+    print('Cluster Sizes are:', cluster_lens)
+
+    #print(q_clusters)
+    #print(grouped)
+    return (grouped, q_clusters)
+  
+
+"""
+> The objective of this function is to cluster questions using Hierarichal Agglomerative Clustering 
+with Sparsification. 
+
+:HAC with sparsification is a novel approach, which modifies the input similarity matrix before feeding it to the HAC Algorithm
+:To read about this approach in detail, see - 
+https://www.researchgate.net/publication/328034827_Improving_Short_Text_Clustering_by_Similarity_Matrix_Sparsification
+
+:The inputs for the function are number of clusters to form, data embeddings, list of questions & type of sparsification to use (1 or 2)
+:The High level function overview is as follows-
+1. Create a similarity matrix from the data embeddings
+2. Modify the matrix by aij = (sij - ui)/std.dev(i) 
+3. From this matrix, extract all max values of complements
+4. Sort this list
+5. Define threshold as: [nxl/2] for l as: n/C  (n: Total Number of Qs, C: No. of Clusters)
+6. Type 1 uses threshold & Type 2 uses 2*threshold
+7. From list keep top values according to required threshold
+8. All left values are ones to be sparsed, i.e made to be 0
+9. Get indexes of left values in original matrix & make them 0
+10.Feed this final matrix into HAC algorithm to create clusters
+"""
+
+
+def HAC_with_Sparsification(n_clus, embeddings, data_used_for_demo, type_n = 1, return_mat = 0):
+  number_of_qs = embeddings.shape[0] #Total number of questions
+
+  #creating initial similarity matrix from data embeddings
+
+  initial_sim_matrix = metrics.pairwise.cosine_similarity(embeddings)
+  mean_list = [] #list to store means of rows
+  std_list = [] #list to store std deviation of rows
+
+  for i in range(len(initial_sim_matrix)):
+    mean_list.append(np.mean(np.concatenate((initial_sim_matrix[i][:i],initial_sim_matrix[i][i+1:]))))
+    std_list.append(np.std(np.concatenate((initial_sim_matrix[i][:i],initial_sim_matrix[i][i+1:]))))
+
+  modified_sim = initial_sim_matrix.copy()
+
+  #creating the modified similarity matrix
+
+  for i in range(len(initial_sim_matrix)):
+    modified_sim[i] = (initial_sim_matrix[i] - mean_list[i])/std_list[i]
+
+  #list to store max of values of all complements
+  L = []
+  for i in range(len(modified_sim)):
+    for j in range(i+1, len(modified_sim)):
+      #j>i guaranteed
+      L.append((max(modified_sim[i][j],modified_sim[j][i]),initial_sim_matrix[i][j]))
+
+  #Now to sort the list
+  L_mod = []
+  for i,j in enumerate(L): #enumerating to have indexes in the list
+    L_mod.append((i+1,j))
+
+  L_mod.sort(key = lambda x: x[1][0], reverse = True)
+
+  #Defining the Threshold
+
+  #threshold = int((number_of_qs*((number_of_qs/n_clus)-1))/2) Can have n/C - 1 too
+  threshold = int((number_of_qs*((number_of_qs/n_clus)))/2)
+
+  #Creating the type 1 (M threshold) Similarity Matrix
+  thresholded_list = sorted(L_mod[threshold:],key = lambda x: x[1][1], reverse=True)
```

Comment:
I think we probably want to move these operations under the conditional checking type_n below so that we only compute the threshold_list that we're actually going to use.